File: go/luhn/luhn.go
package luhn

import (
	"unicode"
)

func Valid(luhnStr string) bool {
	luhnRune := []rune(luhnStr)

	var pos, sum int
	for i := len(luhnRune) - 1; i >= 0; i-- {
		if unicode.IsSpace(luhnRune[i]) {
			continue
		}

		if !unicode.IsDigit(luhnRune[i]) {
			return false
		}

		numeric := int(luhnRune[i]) - '0'
		if pos%2 == 0 {
			sum += numeric
		} else {
			sum += numeric * 2
			if numeric > 4 {
				sum -= 9
			}
		}
		pos++
	}

	if pos < 2 {
		return false
	}
	return sum%10 == 0
}

File: cpp/luhn/luhn.cpp
#include "luhn.h"
#include <cctype>
#include <numeric>

using namespace std;

bool is_digit(char c) {
    auto uc = static_cast<unsigned char>(c);
    return isdigit(uc);
}

int double_int(int num) {
    int doubled = num * 2;
    return (doubled > 9) ? doubled - 9 : doubled;
}

bool luhn::valid(string_view input) {
    int position = 0;
    int luhn_sum = 0;

    for (auto c_itr = input.crbegin(); c_itr != input.crend(); c_itr++) {
        auto c = *c_itr;
        if (is_digit(c)) {
            ++position;
            auto value = c - '0';
            luhn_sum += (position % 2 == 0) ? double_int(value) : value;
        } else if (c != ' ') {
            return false;
        }
    }

    return luhn_sum % 10 == 0 && position > 1;
}
